File: backend/internal/application/usecase/url.savior.usecase.go
package usecase

import (
	"github.com/marcosfrancomarinho/short_url/internal/application/dto"
	"github.com/marcosfrancomarinho/short_url/internal/domain/entities"
	"github.com/marcosfrancomarinho/short_url/internal/domain/gateway"
	"github.com/marcosfrancomarinho/short_url/internal/domain/valuesobject"
)

type URLSaviorUseCase struct {
	idGenerator      gateway.IdGenerator
	urlSavior        gateway.URLSavior
	shortIdGenerator gateway.ShortIdGenerator
	qrCodeGenerator  gateway.QrCodeGenerator
}

func NewURLSaviorUseCase(
	idGenerator gateway.IdGenerator,
	urlSavior gateway.URLSavior,
	shortIdGenerator gateway.ShortIdGenerator,
	qrCodeGenerator gateway.QrCodeGenerator,
) *URLSaviorUseCase {
	return &URLSaviorUseCase{idGenerator: idGenerator, urlSavior: urlSavior, shortIdGenerator: shortIdGenerator, qrCodeGenerator: qrCodeGenerator}
}

func (u *URLSaviorUseCase) Save(paylod *dto.RequestURLSaviorDTO) (*dto.ResponseURLSaviorDTO, error) {
	id, err := u.idGenerator.Generate()
	if err != nil {
		return nil, err
	}

	shortId, err := u.shortIdGenerator.Generator()
	if err != nil {
		return nil, err
	}

	longUrl, err := valuesobject.NewLongUrl(paylod.LongUrl)
	if err != nil {
		return nil, err
	}

	url := entities.NewURL(id, shortId, longUrl)

	if err := u.urlSavior.Save(url); err != nil {
		return nil, err
	}

	shortUrl := url.GetShortUrl(paylod.Host)

	qrCode, err := u.qrCodeGenerator.Generate(url)
	if err != nil {
		return nil, err
	}

	return &dto.ResponseURLSaviorDTO{
		UrlShort: shortUrl,
		ID:       url.GetID(),
		QrCode:   qrCode.GetQrCode(),
	}, nil
}

File: frontend/src/application/usecase/url.savior.usecase.ts
import { URL } from '../../domain/entities/url';
import type { HttpPostClient } from '../../domain/gateway/http.post.client';
import type { RequestUrlSaviorDTO } from '../dto/request.url.savior.dto';
import type { ResponseUrlSaviorDTO } from '../dto/response.url.savior.dto';

export class UrlSaviorUseCase {
  private httpPostClient: HttpPostClient;
  public constructor(httpPostClient: HttpPostClient) {
    this.httpPostClient = httpPostClient;
  }

  public async save(payload: RequestUrlSaviorDTO): Promise<ResponseUrlSaviorDTO> {
    const url: URL = new URL(payload.longUrl);
    console.log(url)
    const response = await this.httpPostClient.execute<ResponseUrlSaviorDTO>(url);
    return response;
  }
}
